### DISCUSSION TITLE
Caching dynamically generated pages

### DISCUSSION THREAD

**Discussion by CaptainT33mo** [vercel/next.js#12245] [Help]
Hi,

I would like to know id there a way to cache the dynamically generated pages on the server for the first load and return the page from cache for the next subsequent loads?

I am using `getInitialProps` in my code to fetch the data from API and generate the pages and then return them. 

Also, I am currently hosting the site on GCP with Ngnix and PM2 so any help in the caching here will be much appreciated.

I would also like to know how the caching can be invalidated if the data changes in the API for that page. SSG is not an option for us since we only have like around 25 pages statically generated and over 100K (and counting) pages generated from the server-side based on the APIs. So any help in this department would be much appreciated.

**Comment by brianlovin** [comment]
I'm wondering if incremental SSG will be a good solution for you, when https://github.com/zeit/next.js/discussions/11552 lands. You could basically SSG a subset of your 100k+ dynamic pages, provide the `fallback: true` flag to incrementally SSG pages as they are visited, and then supply the `revalidate` flag to set some kind of duration where you'd be comfortable serving stale data.

**Reply by CaptainT33mo** [reply]
@brianlovin 
How to SSG a subset of those pages? Like if each of my pages are generated from SSR via API response then how will SSG work here? And what technique are you suggesting to pick those subsets of pages for SSG?

And will those routes which were hit for the first time will they be caches on the server for the next time if someone else hits that route?

**Reply by brianlovin** [reply]
You can define a subset of pages using the `getStaticPaths` method, which allows you to define a list of pages to be created at build time. You can return `fallback: true` to tell Vercel that for non-build-time pages, when they are visited for the _first_ time, to build and cache that page for future visitors.

More info on this here: https://nextjs.org/docs/basic-features/data-fetching#getstaticpaths-static-generation

For defining the subset of pages, that's up to you. You could imagine, for example, pre-building highly trafficked pages, or any pages where speed are highly critical (without that much dynamic data changing between page views).

**Reply by CaptainT33mo** [reply]
But I am not using Vercel's hosting. I am hosting my website on GCP so will that the same ways as well?

**Reply by brianlovin** [reply]
I'm not sure exactly how to handle this outside of Vercel to be honest, maybe someone else can chime in here :)

**Reply by timneutkens** [reply]
`next start` supports fallback: true too, main difference is that on [Vercel](https://vercel.com) those pages will be served from the edge network (CDN) automatically.

**Reply by CaptainT33mo** [reply]
Great. Thanks @brianlovin and @timneutkens . I am running `next build` and `next start` on my server and I'll definitely explore the `getStaticPaths`.
